### PR TITLE
Fix empty tensor for unique_dim

### LINKED ISSUE #18408: _unique_dim does not support zero size tensor
Unique without dim works well
```python
>>> a = torch.empty(5, 0)

>>> a
tensor([], size=(5, 0))

>>> a.unique(return_inverse=True)
(tensor([]), tensor([], size=(5, 0), dtype=torch.int64))
```

But unique with dim does not work
```python
>>> a.unique(dim=1, return_inverse=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/gaoxiang/anaconda3/lib/python3.7/site-packages/torch/tensor.py", line 351, in unique
    dim=dim
RuntimeError: cannot reshape tensor of 0 elements into shape [0, -1]

>>> a.unique(dim=0, return_inverse=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/gaoxiang/anaconda3/lib/python3.7/site-packages/torch/tensor.py", line 351, in unique
    dim=dim
RuntimeError: cannot reshape tensor of 0 elements into shape [-1, 0]
```

I suggest changing the behavior to:
if `dim` refers to a dimension with size 0, and other dimensions have size > 0, then the output and reverse_indicies should be the same empty tensor as input. If `dim` refers to a dimension with size > 0, or there are other 0 sized dimensions, then reject the input with an error because whether "nothing == nothing" is not defined.

### PR DESCRIPTION
**Pull request by codexetreme**
Fixes: #18408

cc: @zasdfgbnm

### CODE REVIEW

File: aten/src/ATen/native/Unique.cpp [hunk 1]
```diff
@@ -78,6 +78,26 @@ std::tuple<Tensor, Tensor> _unique_dim_cpu_template(
...
+
+  // check if dim is in range and see if the size along that dim is 0    
+  if (dim <= self.dim() && self.size(dim)==0 ){
+    // check the other remaining dimentions, returns true if other dimentions are 0
+    bool well_formed_tensor = false;
+    for (int64t i=0; i< self.dim(); ++i){
+      if (dim == i) continue;
+      if (self.size(i) != 0){
+        well_formed_tensor = true;
+        break;
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/Unique.cpp]: I don't think this works either. You need to make sure that all dimensions are nonzero, not just one of it.

File: aten/src/ATen/native/Unique.cpp [hunk 2]
```diff
@@ -78,6 +78,26 @@ std::tuple<Tensor, Tensor> _unique_dim_cpu_template(
     const Tensor& self,
     const int64_t dim,
     const bool return_inverse) {
+
+  // check if dim is in range and see if the size along that dim is 0    
+  if (dim <= self.dim() && self.size(dim)==0 ){
+    // check the other remaining dimentions, returns true if other dimentions are 0
+    bool well_formed_tensor = false;
+    for (int64t i=0; i< self.dim(); ++i){
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/Unique.cpp]: Should be `int64_t`.

**Review by zasdfgbnm — CHANGES_REQUESTED** [review]
Hi @codexetreme, could you please complete this PR first, i.e. to add unit test, and do the same check for GPU? By adding unit test, you could basically find problem by yourself, and change it without having to ask help. The error messages in circle CI should be fairly helpful. Also, please resolve merge conflicts.

File: aten/src/ATen/native/cuda/Unique.cu [hunk 3]
```diff
@@ -67,6 +67,25 @@ template <typename scalar_t>
...
+          well_formed_tensor = false;
+          break;
+        }
+    }
+    if (well_formed_tensor) {
+      Tensor output = self;
+      Tensor inverse_indices =
+          at::empty(self.sizes(), self.options().dtype(kLong));
+          
+      THCudaCheck(cudaGetLastError());
```
Comment (codexetreme) [inline aten/src/ATen/native/cuda/Unique.cu]: Here, I am not sure if the cuda code is right. Could you please check and let me know what has to be done?

File: aten/src/ATen/native/Unique.cpp (hunk 1 above)
Comment (codexetreme) [reply aten/src/ATen/native/Unique.cpp]: I have fixed this, hopefully, this is what is required?

File: aten/src/ATen/native/Unique.cpp (hunk 2 above)
Comment (codexetreme) [reply aten/src/ATen/native/Unique.cpp]: Ah yes, sorry my bad. I have fixed this.

**Review by codexetreme — COMMENTED** [review]
I have fixed merge conflicts, the other irregularities you mentioned. Also I am having a bit of difficulty writing tests. Where are these supposed to go exactly? there are many test folders. Should I add the tests in `test_torch.py` or in some other test folder?

**Comment by zasdfgbnm** [comment]
@codexetreme Sorry for the late reply, yes, for the test, you need to write it at `test_torch.py`. Just search for `test_unique_dim` in that file. I will check your work later. But in the mean time, could you resolve the merge conflicts again?

**Comment by zasdfgbnm** [comment]
BTW, I am working on unique also, so some of my work might frequently cause merge conflicts for this. You might also continue your work, and resolve the conflicts later, which might be of less trouble. It's your choice.

File: aten/src/ATen/native/cuda/Unique.cu (hunk 3 above)
Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Unique.cu]: Well, the logic for checking shape should be the same for CPU and CUDA, correct? In this case, why not create a `Unique.h` file, put the logic there as a function, and just call that function for CPU and CUDA?

**Comment by zasdfgbnm** [comment]
@codexetreme Also, whenever you see build failures from CI, could you also please check where is error and try fixing it? :)

**Comment by codexetreme** [comment]
Alright, I am looking into the CI logs... I'll write the tests as well.

> You might also continue your work, and resolve the conflicts later, which might be of less trouble.

I will do this..

**Comment by codexetreme** [comment]
So in some of the CI logs, the error that is said is :
>Reading package lists...
Building dependency tree...
Reading state information...
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

>The following packages have unmet dependencies:
 nvidia-docker2 : Depends: nvidia-container-runtime (= 2.0.0+docker18.06.2-1) but 2.0.0+docker18.06.2-2 is to be installed
E: Unable to correct problems, you have held broken packages.
Exited with code 100

Does this mean that the containers have not been properly instantiated or that the code in the commit is broken? And what to do for these?

**Comment by zasdfgbnm** [comment]
@codexetreme For these kind of failures, you could do a "@pytorchbot retest this please"

**Comment by zasdfgbnm** [comment]
@codexetreme The bot is currently broken. You may try again next week.

**Comment by codexetreme** [comment]
The same issue persists with the tests. What should be done ?

**Comment by zasdfgbnm** [comment]
@codexetreme I think you need to resolve the conflict and merge with the latest master branch now. They must have updated CircleCI configurations.

**Comment by codexetreme** [comment]
Sorry for the late reply (uni assignments take up time :(  ). Could you tell me the use of the `counts` bool? my code is failing as I am not returning a third tensor that should contain the counts tensor. According to my understanding, if the tensor is empty should I just return an empty tensor for it? (since there are no elements in the tensor and count would be zero anyway)

**Comment by zasdfgbnm** [comment]
@codexetreme Yes

**Comment by codexetreme** [comment]
Sorry for the late reply, In unique.cu, the error I'm getting is 

> Apr 18 22:10:06 
Apr 18 22:10:06 /var/lib/jenkins/workspace/aten/src/ATen/native/cuda/Unique.cu(167): error: "input_flat" has already been declared in the current scope
Apr 18 22:10:06 

in particular, the error is because input_flat has already been declared previously, a simple fix is to simply rename the variable in one of the 2 places ( and all its corresponding usages).

**Comment by codexetreme** [comment]
In the `Unique.cu` file, in `unique_dim_cuda_template` `input_flat`, is declared twice, a simple fix is to just rename its usages here, since I dont see these particular lines being used elsewhere.
`Tensor input_flat = self.transpose(dim, 0);`
`auto orig_sizes = input_flat.sizes().vec();`
 `input_flat = input_flat.contiguous().view({input_flat.size(0), -1});`

**Comment by codexetreme** [comment]
Hello, sorry to bother you again, but, I wanted to know if I can run just 1 test (any CPU based build and test is fine) because I cannot compile the code on my system, and so I will have to keep committing and wait till all the CI's are run and then check if tests pass and/or are made correctly. If I can download the python wheels from this current push that also is fine so that I can check the tests locally. Thanks for your patience :)

**Comment by codexetreme** [comment]
So all checks have passed,  whatv to do next?

File: test/test_torch.py [hunk 4]
```diff
@@ -10832,7 +10833,9 @@ def run_test(dtype=torch.float, device=torch.device('cpu')):
                                                 device=device)
             expected_inverse_dim2 = torch.tensor([0, 1])
             expected_counts_dim2 = torch.tensor([1, 1])
-
+            expected_inverse_empty = torch.zeros(5, 0, dtype=dtype, device=device)
```
Comment (zasdfgbnm) [inline test/test_torch.py]: Shouldn't inverse index always be long tensor?

File: test/test_torch.py [hunk 5]
```diff
@@ -10832,7 +10833,9 @@ def run_test(dtype=torch.float, device=torch.device('cpu')):
...
+            expected_unique_empty = torch.tensor([], dtype=dtype, device=device)
+            expected_counts_empty = torch.tensor([], dtype=dtype, device=device)
```
Comment (zasdfgbnm) [inline test/test_torch.py]: Same here

File: aten/src/ATen/native/Unique.cpp [hunk 6]
```diff
@@ -154,6 +154,28 @@ std::tuple<Tensor, Tensor, Tensor> _unique_dim_cpu_template(
...
     const bool return_counts) {
+  // check if dim is in range and see if the size along that dim is 0
+  if (dim <= self.dim() && self.size(dim) == 0) {
+    // assume tensor has all dimentions non-zero
+    bool well_formed_tensor = true;
+    for (int64_t i = 0; i < self.dim(); ++i) {
+      if (dim == i)
+        continue;
+      if (self.size(i) == 0) {
+        // found atleast 1 zero sized dimention, hence the tensor is not well formed
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/Unique.cpp]: nit: "atleast" and "dimention"

File: test/test_torch.py [hunk 7]
```diff
@@ -10917,6 +10920,16 @@ def run_test(dtype=torch.float, device=torch.device('cpu')):
             self.assertEqual(expected_inverse_dim2, x_inverse)
             self.assertEqual(expected_counts_dim2, x_counts)
 
+            # test empty tensor
```
Comment (zasdfgbnm) [inline test/test_torch.py]: Could you also add the test case when the tensor is not well formed?

File: aten/src/ATen/native/Unique.cpp [hunk 8]
```diff
@@ -154,6 +154,28 @@ std::tuple<Tensor, Tensor, Tensor> _unique_dim_cpu_template(
...
+        well_formed_tensor = false;
+        break;
+      }
+    }
+    if (well_formed_tensor) {
+      Tensor output = at::empty({0}, self.options().dtype(kLong));
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/Unique.cpp]: Why output is dtype long, instead of the same type as self?

File: aten/src/ATen/native/Unique.cpp [hunk 9]
```diff
@@ -154,6 +154,28 @@ std::tuple<Tensor, Tensor, Tensor> _unique_dim_cpu_template(
...
+      Tensor inverse_indices =
+          at::empty(self.sizes(), self.options().dtype(kLong));
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/Unique.cpp]: For inverse indices of `unique_dim`, shouldn't it be a 1D vector?

File: aten/src/ATen/native/cuda/Unique.cu [hunk 10]
```diff
@@ -138,6 +138,30 @@ std::tuple<Tensor, Tensor, Tensor> unique_dim_cuda_template(
...
+          break;
+        }
+    }
+    if (well_formed_tensor) {
+      Tensor output = at::empty({0},self.options().dtype(kLong));
+      Tensor inverse_indices =
+          at::empty(self.sizes(), self.options().dtype(kLong));
+      Tensor counts = at::empty({0}, self.options().dtype(kLong));
+          
+      THCudaCheck(cudaGetLastError());
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/Unique.cu]: I don't think there is need to do `THCudaCheck`

File: aten/src/ATen/native/cuda/Unique.cu [hunk 11]
```diff
@@ -138,6 +138,30 @@ std::tuple<Tensor, Tensor, Tensor> unique_dim_cuda_template(
...
+        if (dim == i)
+          continue;
+        if (self.size(i) == 0) {
+          // found atleast 1 zero sized dimention, hence the tensor is not well formed
+          well_formed_tensor = false;
+          break;
+        }
+    }
+    if (well_formed_tensor) {
+      Tensor output = at::empty({0},self.options().dtype(kLong));
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/Unique.cu]: Same problem here as in the review of CPU version

File: aten/src/ATen/native/cuda/Unique.cu [hunk 12]
```diff
@@ -138,6 +138,30 @@ std::tuple<Tensor, Tensor, Tensor> unique_dim_cuda_template(
...
+      Tensor inverse_indices =
+          at::empty(self.sizes(), self.options().dtype(kLong));
+      Tensor counts = at::empty({0}, self.options().dtype(kLong));
+          
+      THCudaCheck(cudaGetLastError());
+      return std::make_tuple(output, inverse_indices,counts);
+    }
+  }
+    Tensor input_flat_ = self.transpose(dim, 0);
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/Unique.cu]: What are these lines for?

**Review by zasdfgbnm — CHANGES_REQUESTED** [review]
@codexetreme Please see the review

**Comment by codexetreme** [comment]
Hello, I have fixed the changes you specified. these 4 tests are pending for long, so I am sure about their status. Any other fixes are required?

File: aten/src/ATen/native/Unique.cpp [hunk 13]
```diff
@@ -154,6 +154,28 @@ std::tuple<Tensor, Tensor, Tensor> _unique_dim_cpu_template(
...
     const bool return_inverse,
     const bool return_counts) {
+  // check if dim is in range and see if the size along that dim is 0
+  if (dim <= self.dim() && self.size(dim) == 0) {
+    // assume tensor has all dimentions non-zero
+    bool well_formed_tensor = true;
+    for (int64_t i = 0; i < self.dim(); ++i) {
+      if (dim == i)
+        continue;
+      if (self.size(i) == 0) {
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/Unique.cpp]: Would it be better to change the whole `if` to something like
```C++
AT_CHECK(self.size(i) > 0, "Some error message here");
```

File: aten/src/ATen/native/Unique.cpp [hunk 14]
```diff
@@ -154,6 +154,28 @@ std::tuple<Tensor, Tensor, Tensor> _unique_dim_cpu_template(
     const bool consecutive,
     const bool return_inverse,
     const bool return_counts) {
+  // check if dim is in range and see if the size along that dim is 0
+  if (dim <= self.dim() && self.size(dim) == 0) {
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/Unique.cpp]: What would happen if `self.size(dim) != 0` but there is another dim that has zero size? Shouldn't this case be handled as well?

File: aten/src/ATen/native/cuda/Unique.cu [hunk 15]
```diff
@@ -138,6 +138,27 @@ std::tuple<Tensor, Tensor, Tensor> unique_dim_cuda_template(
   cudaStream_t stream = at::cuda::getCurrentCUDAStream();
   auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
   auto policy = thrust::cuda::par(allocator).on(stream);
+  
+  if (dim <= self.dim() && self.size(dim) == 0) {
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/Unique.cu]: Please fix this part following the review of CPU implementation accordingly :)

File: aten/src/ATen/native/Unique.cpp [hunk 16]
```diff
@@ -154,6 +154,28 @@ std::tuple<Tensor, Tensor, Tensor> _unique_dim_cpu_template(
...
+      if (self.size(i) == 0) {
+        // found at least 1 zero sized dimension, hence the tensor is not well formed
+        well_formed_tensor = false;
+        break;
+      }
+    }
+    if (well_formed_tensor) {
+      Tensor output = at::empty({0}, self.options());
+      Tensor inverse_indices =
+          at::empty(self.contiguous().view({self.size(0),-1}).sizes(), self.options().dtype(kLong));
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/Unique.cpp]: Shouldn't the shape of inverse indices be `{self.size(dim)}`?

**Review by zasdfgbnm — CHANGES_REQUESTED** [review]
@codexetreme Please see the comments

File: test/test_torch.py [hunk 17]
```diff
@@ -7625,6 +7625,11 @@ def test_masked_scatter(self):
             with self.assertRaises(RuntimeError):
                 dest.masked_scatter_(mask, src)
 
+            # make src smaller. this should fail
```
[the remark this answers is not in the document]
Comment (codexetreme) [reply test/test_torch.py]: this was due to a merge conflict, i'm not sure what why it gave one there, but i selected the lines  that the master branch has, ill roll back and fix it

File: aten/src/ATen/native/Unique.cpp (hunk 13 above)
Comment (codexetreme) [reply aten/src/ATen/native/Unique.cpp]: is there any documentation on such functionality? I feel like I'm missing out on a lot of these predefined functions/macros

File: aten/src/ATen/native/Unique.cpp (hunk 16 above)
Comment (codexetreme) [reply aten/src/ATen/native/Unique.cpp]: Oh, I thought it returned all indices, my bad, I will fix it

**Comment by codexetreme** [comment]
So, I have made the changes you requested, I changed the logic so now it handles the `self.size(dim) !=0` case as well. Can you please check and let me know it is right?

File: aten/src/ATen/native/Unique.cpp [hunk 18]
```diff
@@ -154,6 +154,28 @@ std::tuple<Tensor, Tensor, Tensor> _unique_dim_cpu_template(
...
+    auto num_zero_dims = std::count(sizes.begin(), sizes.end(),0);
+    
+    // tensor is not well formed as it has 0 sized dimentions
+    if (self.size(dim)==0){
+      AT_CHECK(
+          num_zero_dims == 1,
+          "Number of zero sized dimentions is more than one, so unique cannot be applied ")
+      Tensor output = at::empty({0}, self.options());
+      Tensor inverse_indices =
+          at::empty(self.size(dim), self.options().dtype(kLong));
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/Unique.cpp]: Isn't `size(dim)` just 0?

File: aten/src/ATen/native/Unique.cpp [hunk 19]
```diff
@@ -154,6 +154,28 @@ std::tuple<Tensor, Tensor, Tensor> _unique_dim_cpu_template(
...
+      Tensor counts = at::empty({0}, self.options().dtype(kLong));
+
+      return std::make_tuple(output, inverse_indices, counts);
+    }
+    if (self.size(dim) != 0 ){
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/Unique.cpp]: Isn't `self.size(dim) != 0` always true here?

**Review by zasdfgbnm — COMMENTED** [review]
It looks much better now. Could you please also fix the linter failures in Travis CI?

```
./test/test_torch.py:10963:17: E128 continuation line under-indented for visual indent
./test/test_torch.py:10964:17: E128 continuation line under-indented for visual indent
./test/test_torch.py:10965:17: E128 continuation line under-indented for visual indent
./test/test_torch.py:10966:17: E128 continuation line under-indented for visual indent
./test/test_torch.py:10969:17: E128 continuation line under-indented for visual indent
./test/test_torch.py:10970:17: E128 continuation line under-indented for visual indent
./test/test_torch.py:10971:17: E128 continuation line under-indented for visual indent
./test/test_torch.py:10972:17: E128 continuation line under-indented for visual indent
The command "flake8" exited with 1.
```

File: aten/src/ATen/native/Unique.cpp (hunk 13 above)
Comment (zasdfgbnm) [reply aten/src/ATen/native/Unique.cpp]: I don't think there is any document for these functions. The best way to learn may be looking at other codes that uses those functions...

File: aten/src/ATen/native/Unique.cpp (hunk 18 above)
Comment (codexetreme) [reply aten/src/ATen/native/Unique.cpp]: yes, I will change it

**Comment by codexetreme** [comment]
Ok, so I have fixed the linting

File: aten/src/ATen/native/Unique.cpp [hunk 20]
```diff
@@ -154,6 +154,28 @@ std::tuple<Tensor, Tensor, Tensor> _unique_dim_cpu_template(
     const bool consecutive,
     const bool return_inverse,
     const bool return_counts) {
+
+    auto sizes = self.sizes().vec();
+    // check how many zero dimentions exist
+    auto num_zero_dims = std::count(sizes.begin(), sizes.end(),0);
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/Unique.cpp]: nit: space between `sizes.end(),` and `0`

File: aten/src/ATen/native/Unique.cpp [hunk 21]
```diff
@@ -154,6 +154,28 @@ std::tuple<Tensor, Tensor, Tensor> _unique_dim_cpu_template(
...
+    
+    // tensor is not well formed as it has 0 sized dimentions
+    if (self.size(dim)==0){
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/Unique.cpp]: nit: space before and after `==`

File: aten/src/ATen/native/Unique.cpp (hunk 19 above)
Comment (zasdfgbnm) [reply aten/src/ATen/native/Unique.cpp]: @codexetreme How about this? You are basically checking a condition that is always true.

Comment (codexetreme) [reply aten/src/ATen/native/Unique.cpp]: yep, you're right, I've kept it there to check the condition where the selected dim is non zero in size and there are zero sized dims in the tensor, but in hindsight, it's not going to do much there, ill remove it. :)

**Comment by codexetreme** [comment]
Why is that 1 particular test failing? The other cuda based tests seem to have passed...

**Review by zasdfgbnm — APPROVED** [review]
@codexetreme That failure is not your fault, just ignore it. The code LGTM now. @VitalyFedyunin Could you please take a look and merge?

**Comment by codexetreme** [comment]
When will this be merged? Kinda curious .....

**Comment by zasdfgbnm** [comment]
kindly ping @VitalyFedyunin :)

**Comment by VitalyFedyunin** [comment]
Sure, I was hoping to have #20017  fixed (this helps to write tests quickly), guess I will do workaround.

File: aten/src/ATen/native/Unique.cpp [hunk 22]
```diff
@@ -154,6 +154,27 @@ std::tuple<Tensor, Tensor, Tensor> _unique_dim_cpu_template(
     const bool consecutive,
     const bool return_inverse,
     const bool return_counts) {
+
+    auto sizes = self.sizes().vec();
+    // check how many zero dimentions exist
+    auto num_zero_dims = std::count(sizes.begin(), sizes.end(), 0);
+    
+    // tensor is not well formed as it has 0 sized dimentions
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/Unique.cpp]: nit: dimenSions, here and everywhere

File: test/test_torch.py [hunk 23]
```diff
@@ -10963,6 +10968,31 @@ def run_test(dtype=torch.float, device=torch.device('cpu')):
...
+                return_inverse=True,
+                return_counts=True,
+                dim=1)
+            self.assertEqual(expected_unique_empty, x_unique)
+            self.assertEqual(expected_inverse_empty, x_inverse)
+            self.assertEqual(expected_counts_empty, x_counts)
+
+            # test not a well formed tensor
+            # Checking for runtime error, as this is the expected behaviour
+            self.assertRaises(
```
Comment (VitalyFedyunin) [inline test/test_torch.py]: nit: It is cleaner to write
```python
with self.assertRaises(RuntimeError):
  torch.unique(
                x_ill_formed_empty, 
                return_inverse=True, 
                return_counts=True, 
                dim=1)
```

**Review by VitalyFedyunin — APPROVED** [review]

File: test/test_torch.py (hunk 23 above)
Comment (codexetreme) [reply test/test_torch.py]: I have made the change like suggested :)

File: aten/src/ATen/native/Unique.cpp (hunk 22 above)
Comment (codexetreme) [reply aten/src/ATen/native/Unique.cpp]: fixed this :)

**Comment by VitalyFedyunin** [comment]
Could you please cleanup training whitespaces

**Pull request closed without merging**